Reference documentation for the Window Controls Overlay API

captainbrosset opened the pull request:
<!-- 👀 Thanks for opening a PR! Read comments like this one to get your PR merged faster. -->
#### Summary
This PR adds the required documentation for the WindowControlsOverlay JS API.

#### Motivation
Desktop-installed PWAs can now use the Window Controls Overlay feature (available generally starting from Chrome/Edge 98).

Using this feature requires using the `window-controls-overlay` display mode in the `display_override` member, which was already documented here: https://developer.mozilla.org/en-US/docs/Web/Manifest/display_override

And the feature also relies on new CSS environment variables (`titlebar-area-*`) which were also already documented here: https://developer.mozilla.org/en-US/docs/Web/CSS/env()#using_env_to_ensure_content_is_not_obscured_by_window_control_buttons_in_desktop_pwas

The remaining piece for this feature is a JavaScript API: `navigator.windowControlsOverlay` which I am documenting in this PR.
The API isn't very complicated, it has just one property (`visible`), one event handler (`ongeometrychange`), and one method (`getTitlebarAreaRect`).

I believe I have made all of the necessary changes for these things to be documented.

**Notes**

* I did not create a separate interface reference page for the WindowControlsOverlayGeometryChangeEvent. This event is (for now) simple enough that I documented its properties in the event page, instead of creating another one for the interface.
* The version of BCD currently in use needs to be updated to fix the current BCD errors. I am adding the necessary compat data in the BCD repo here: https://github.com/mdn/browser-compat-data/pull/14095.

#### Related issues
This is part of this content proposal parent issue: https://github.com/mdn/content/issues/10539

#### Metadata
- [x] Adds a new document
- [x] Rewrites (or significantly expands) a document
- [ ] Fixes a typo, bug, or other error

<!-- 👷‍♀️ After submitting, review the results of the "Checks" tab! -->

captainbrosset committed:
Initial content for the WCO API documentation

captainbrosset committed:
Fix the BCD references

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/geometrychange_event/index.md:
```diff
@@ -0,0 +1,72 @@
+
+The `geometrychange` event is fired when the position, size, or visibility of a desktop Progressive Web App's title bar area changes.
+
+<table class="properties">
+  <tbody>
+    <tr>
+      <th scope="row">Bubbles</th>
+      <td>No</td>
+    </tr>
+    <tr>
+      <th scope="row">Cancelable</th>
+      <td>No</td>
+    </tr>
+    <tr>
+      <th scope="row">Event handler property</th>
+      <td>
+        <code
+          ><a href="/en-US/docs/Web/API/WindowControlsOverlay/ongeometrychange"
+            >ongeometrychange</a
+          ></code
+        >
+      </td>
+    </tr>
+  </tbody>
+</table>
```
```suggestion
## Syntax

Use the event name in methods like {{domxref("EventTarget.addEventListener", "addEventListener()")}}, or set an event handler property.

```js
addEventListener('geometrychange', (event) => { });

geometrychange = (event) => { });
```

```

jpmedley replied:
Don't commit this until we nail down what the event name is going to be. We may need to change it.

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/geometrychange_event/index.md:
```diff
@@ -0,0 +1,72 @@
+  </tbody>
+</table>
+
+The `geometrychange` event has the following properties:
```
```suggestion
## Event type

A {{domxref("GeometryChangeEvent")}}. Inherits from {{domxref("[Event")}}.

{{InheritanceDiagram(500, 70, 50, "GeometryChangeEvent")}}

## Event properties

_In addition to the properties listed below, properties from the parent interface, {{domxref("Event")}}, are available._
```

captainbrosset replied:
@jpmedley I'm assuming that means I need to create a GeometryChangeEvent interface doc as well, is that right?

jpmedley replied:
Yes.

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/geometrychange_event/index.md:
```diff
@@ -0,0 +1,72 @@
+The `geometrychange` event has the following properties:
+
+- `titlebarAreaRect`
+  - : A {{domxref("DOMRect")}} representing the position and size of the title bar area.
```
```suggestion
  - : A {{domxref("DOMRect")}} representing the position and size of the title bar.
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/geometrychange_event/index.md:
```diff
@@ -0,0 +1,72 @@
+---
+title: 'WindowControlsOverlay: geometrychange event'
```
@tomayac Chrome's IDL says `WindowControlsOverlayGeometryChangeEvent` (as does your article). I'm guessing one of these is out of date, but which?

captainbrosset replied:
@diekus could you shade some light here please?

jpmedley replied:
Pinging this thread so we can move this forward. We appreciate the help you folks can give.

captainbrosset replied:
The event name you are supposed to use in JS to listen to WCO geometry changes is: geometrychange. Example:

```javascript
navigator.windowControlsOverlay.addEventListener('geometrychange', () => {});
```

or 

```javascript
navigator.windowControlsOverlay.ongeometrychange = () => {};
```

Which is why I named this page `geometrychange_event/index.md`. Can you explain where `WindowControlsOverlayGeometryChangeEvent` would be used on this page?

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/gettitlebararearect/index.md:
```diff
@@ -0,0 +1,32 @@
+---
+{{APIRef("Window Controls Overlay API")}}{{SeeCompatTable}}
+
+The **`WindowControlsOverlay.prototype.getTitleBarAreaRect()`** method queries the current geometry of the title bar area of the desktop Progressive Web App window.
```
```suggestion
The **`getTitleBarAreaRect()`** method of the {{domxref('WindowControlsOverlay')}} interface queries the current geometry of the title bar area of a desktop Progressive Web App window.
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/gettitlebararearect/index.md:
````diff
@@ -0,0 +1,32 @@
+## Syntax
+
+```js
+navigator.windowControlsOverlay.getTitleBarAreaRect();
````
```suggestion
getTitleBarAreaRect();
```

jpmedley replied:
Per the [current instructions](https://developer.mozilla.org/en-US/docs/MDN/Structures/Syntax_sections#constructors_and_methods).

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/index.md:
```diff
@@ -0,0 +1,52 @@
+
+- {{domxref("WindowControlsOverlay.getTitlebarAreaRect()")}}
+  - : Returns the size and position of the title bar area.
+
```
```suggestion
## Events
Listen to these events using {{domxref('EventTarget.addEventListener()')}} or by assigning an event listener to the relevant `oneventname` property of this interface.

- [`geometrychange`](/en-US/docs/Web/API/VisualViewport/geometrychange_event)
  - : Fired when the geometry of the title bar area changes.
  
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/index.md:
```diff
@@ -0,0 +1,52 @@
+- {{domxref("WindowControlsOverlay.visible")}} {{readonlyInline}}
+  - : A {{Glossary("Boolean")}} that indicates whether the window controls overlay is visible or not.
+
+### Event handlers
+
+- {{domxref("WindowControlsOverlay.onGeometryChange")}}
+  - : A handler for the {{domxref("WindowControlsOverlay/geometrychange_event", "geometrychange")}} event. This event is sent when the geometry of the title bar area changes.
+
```
```suggestion
```

jpmedley replied:
This is moved down and done in a different way.

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/index.md:
```diff
@@ -0,0 +1,52 @@
+## Methods
+
+- {{domxref("WindowControlsOverlay.getTitlebarAreaRect()")}}
+  - : Returns the size and position of the title bar area.
```
```suggestion
  - : Returns the size and position of the title bar.
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/ongeometrychange/index.md:
```diff
@@ -0,0 +1,30 @@
+---
```
Please delete this whole page.

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/visible/index.md:
```diff
@@ -0,0 +1,45 @@
+
+The **`visible`** property of a {{domxref("WindowControlsOverlay")}} object returns a {{Glossary("Boolean")}} that indicates whether the window controls overlay is visible or not.
+
+The window controls overlay will not be visible if the Progressive Web App does not use the Window Controls Overlay feature at all, or if the user has opted-out of the feature.
```
>does not use the Window Controls Overlay feature at all

I would rephrase this in terms of Web App Manifest configuration with a link. The nature of reference is such that you should never assume a reading order.

jpmedley commented on files/en-us/web/css/env()/index.md:
```diff
@@ -172,7 +172,7 @@ The syntax of the fallback, like that of custom properties, allows commas. But,
 ### Using env() to ensure content is not obscured by window control buttons in desktop PWAs
 
-In the following example `env()` is used to ensure that content displayed in a PWA that uses the Window Controls Overlay feature is not obscured by the operating system's window control buttons. The `titlebar-area-*` values define a rectangle where the title bar would normally have been displayed. On devices that do not support the Window Controls Overlay feature, such as mobile devices, the fall ...
+In the following example `env()` is used to ensure that content displayed in a desktop Progressive Web App that uses the [Window Controls Overlay API](/en-US/docs/Web/API/Window_Controls_Overlay_API) is not obscured by the operating system's window control buttons. The `titlebar-area-*` values define a rectangle where the title bar would normally have been displayed. On devices that do not suppor ...
```
```suggestion
In the following example `env()` ensures that content displayed in a desktop Progressive Web App that uses the [Window Controls Overlay API](/en-US/docs/Web/API/Window_Controls_Overlay_API) is not obscured by the operating system's window control buttons. The `titlebar-area-*` values define a rectangle where the title bar would normally have been displayed. On devices that do not support the Window Controls Overlay feature, such as mobile devices, the fallback values are used.
```

jpmedley commented on files/jsondata/GroupData.json:
```diff
@@ -2187,6 +2187,15 @@
         "XRSession: visibilitychange"
       ]
     },
+    "Window Controls Overlay API": {
```
@Elchi3 Can you help me out with a review of GroupData. I'm rusty.

Elchi3 replied:
Looks good to me. 
I mean there isn't anything to group here really, I guess.

tomayac commented on files/en-us/web/api/navigator/windowcontrolsoverlay/index.md:
````diff
@@ -0,0 +1,56 @@
+the title bar area, and know whenever it changes, in order to position web content in
+that area.
+
+## Syntax
+
+```js
+var wco = navigator.windowControlsOverlay
+```
+
+### Value
````
This should also use modern `const` and `let`.

jpmedley replied:
Thanks. MDN decided not to show syntax for properties any more. They just show the return value, which is what developers need to know. Properties all have the same syntax and it's something that doesn't need to be explained.

Elchi3 commented on files/jsondata/GroupData.json:
```diff
@@ -2187,6 +2187,15 @@
+        "WindowControlsOverlay"
+      ],
+      "methods": [],
+      "properties": [],
```
This could be added: 

```suggestion
      "properties": ["Navigator.windowControlsOverlay"],
```

jpmedley replied:
@captainbrosset, please commit this.

hamishwillee commented on files/en-us/web/api/window_controls_overlay_api/index.md:
```diff
@@ -0,0 +1,68 @@
+
+## Specifications
+
+| Specification                         | Status                             | Comment             |
+| ------------------------------------- | ---------------------------------- | ------------------- |
+| {{SpecName("WindowControlsOverlay")}} | {{Spec2("WindowControlsOverlay")}} | Initial definition. |
```
Spec should generally be defined in browser compatibility under the key `api.WindowControlsOverlay` and accessed using `{{Specifications}}` macro. Unless there is a reason you have chosen not to of course :-).

```suggestion
{{Specifications}}
```

captainbrosset committed:
Merge remote-tracking branch 'origin' into windowControlsOverlay

captainbrosset committed:
Addressed review comments

captainbrosset committed:
Merge remote-tracking branch 'origin/main' into windowControlsOverlay

captainbrosset committed:
Added WindowControlsOverlayGeometryChangeEvent

Added short ref doc about this event type.
Also tried to clarify some concepts.
BCD data is missing still, but I have PRs
open for this.

captainbrosset commented:
I just pushed another commit to add the documentation for the `WindowControlsOverlayGeometryChangeEvent` interface.
Some notes about this:

* I previously called it `GeometryChangeEvent`. This was wrong, and I now corrected it to `WindowControlsOverlayGeometryChangeEvent`.
* I tried to clarify some of the previous pages I created too. This concept isn't very straightforward, and so I added a couple sentences on various pages mostly to link people back to the Window Controls Overlay API page when possible. What I feel is the most confusing is that the `titlebarAreaRect` stuff doesn't tell users about the title bar, but rather the area that it would occupy within the web content, if it had not been hidden with the WindowControlsOverlay feature.
* I still have an outstanding BCD pull request that needs to be merged and dependencies to be updated here for some of the spec/compat macros to stop failing: https://github.com/mdn/browser-compat-data/pull/14095

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/titlebararearect/index.md:
````diff
@@ -0,0 +1,32 @@
+**`WindowControlsOverlayGeometryChangeEvent.titlebarAreaRect`** read-only property is a
+{{domxref("DOMRect")}} representing the position and size of the area occupied by the title bar in a desktop-installed Progressive Web App.
+
+## Syntax
+
+```js
+rect = WindowControlsOverlayGeometryChangeEvent.titlebarAreaRect
+```
+
````
```suggestion
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/visible/index.md:
```diff
@@ -0,0 +1,34 @@
+{{APIRef("DOM Events")}}
+
+The
+**`WindowControlsOverlayGeometryChangeEvent.visible`** read-only property is a
```
```suggestion
**`WindowControlsOverlayGeometryChangeEvent.visible`** read-only property of the {{domxref("WindowControlsOverlayGeometryChangeEvent")}} is a
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/visible/index.md:
```diff
@@ -0,0 +1,34 @@
+**`WindowControlsOverlayGeometryChangeEvent.visible`** read-only property is a
+boolean flag that indicates whether the window controls overlay is visible or not in a desktop-installed Progressive Web App.
+
+If the user opts-out of the [Window Controls Overlay](/en-US/docs/Web/API/Window_Controls_Overlay_API) feature, the default title bar appears and the window controls buttons does not appear as an overlay, in which case the `visible` flag is false.
```
```suggestion
If the user opts-out of the [Window Controls Overlay](/en-US/docs/Web/API/Window_Controls_Overlay_API) feature, the default title bar appears and the window controls buttons do not appear as an overlay, in which case the `visible` flag is false.
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/visible/index.md:
````diff
@@ -0,0 +1,34 @@
+
+If the user opts-out of the [Window Controls Overlay](/en-US/docs/Web/API/Window_Controls_Overlay_API) feature, the default title bar appears and the window controls buttons does not appear as an overlay, in which case the `visible` flag is false.
+
+## Syntax
+
+```js
+flag = WindowControlsOverlayGeometryChangeEvent.visible
+```
+
````
```suggestion
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/visible/index.md:
```diff
@@ -0,0 +1,34 @@
+
+{{Compat}}
+
+## See also
+
+- The {{domxref("WindowControlsOverlayGeometryChangeEvent")}} interface it belongs to.
```
```suggestion
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/titlebararearect/index.md:
```diff
@@ -0,0 +1,32 @@
+
+{{Compat}}
+
+## See also
+
+- The {{domxref("WindowControlsOverlayGeometryChangeEvent")}} interface it belongs to.
```
```suggestion
```

jpmedley replied:
Because it's in the intro.

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/windowcontrolsoverlaygeometrychangeevent/index.md:
```diff
@@ -0,0 +1,45 @@
+{{domxref("Event.Event", "Event()")}}._
+
+- _type_
+  - : Is a {{domxref("DOMString")}} representing the name of the type of the
```
```suggestion
  - : A {{domxref("DOMString")}} indicating the event type, always 
```

captainbrosset replied:
I'm not sure I follow what you mean by `always`?
The value of this string is `change` when the event is fired. I don't know if any other string values are permitted when constructing the event though.

captainbrosset replied:
I ended up removing the rest of the sentence, and only keeping `A {{domxref("DOMString")}} indicating the event type.` Is that ok?

jpmedley replied:
What I'm trying to convey is that the exact type name of the event must be supplied in that parameter, and that that name is case sensitive. I also want to save readers from guessing. The spec doesn't make this clear because spec implementors are expected to know the idiom.

jpmedley commented on files/en-us/web/api/windowcontrolsoverlaygeometrychangeevent/windowcontrolsoverlaygeometrychangeevent/index.md:
```diff
@@ -0,0 +1,45 @@
+- _type_
+  - : Is a {{domxref("DOMString")}} representing the name of the type of the
+    `WindowControlsOverlayGeometryChangeEvent`. It is case-sensitive.
+- `visible`
+  - : Is a boolean flag that's true when the titleBarAreaRect's values are not 0.
+- `titleBarAreaRect`
+  - : Is a {{domxref("DOMRect")}} representing the position and size of the title bar area.
```
```suggestion
- `options`
  - : An object with the following properties:
  - `visible`
    - : A boolean flag that's true when the `titleBarAreaRect` object's values are not 0.
  - `titleBarAreaRect`
    - : A {{domxref("DOMRect")}} representing the position and size of the title bar area.
```

jpmedley commented on files/en-us/web/api/windowcontrolsoverlay/gettitlebararearect/index.md:
```diff
@@ -0,0 +1,34 @@
+---
+title: WindowControlsOverlay.prototype.getTitleBarAreaRect()
```
```suggestion
title: WindowControlsOverlay.getTitleBarAreaRect()
```

jpmedley replied:
I apologize for net catching this on the first review. We don't do this any more because 'prototype' wont' appear in a script. If there's a reason to distinguish between static and instance methods on a particular interface it's added to the basic description. For example, "The doStuff() static method of the whatever interface...".

jpmedley reviewed (changes requested):
I appreciate your patience, and thank you for adding the additional pages. I know this is a difficult PR.

captainbrosset committed:
Merge remote-tracking branch 'origin/main' into windowControlsOverlay

captainbrosset committed:
Addressed code review feedback

captainbrosset committed:
Merge remote-tracking branch 'origin/main' into windowControlsOverlay

captainbrosset committed:
Added WCO event name value

captainbrosset commented:
Thanks for the reviews so far @jpmedley. I think I've addressed all comments. Let me know what you think.
Ideally, this PR would be ready before the WCO feature becomes available to all in stable versions of chromium browsers: beginning of February for Chrome and Edge.

jpmedley reviewed (approved):
Thank you for your patience.

captainbrosset commented:
Thanks @jpmedley. Can this be merged then?

captainbrosset commented:
Thank you Joe!

Just for reference, the following 2 BCD PRs have also been merged and we will need to update MDN's dependency on BCD to get the proper spec/compat-data information on the new pages added here:

https://github.com/mdn/browser-compat-data/pull/14035
https://github.com/mdn/browser-compat-data/pull/14095

jpmedley commented:
BCD data is typically deployed once a week. It wasn't after these merges because of the holidays, but there was one yesterday. These should be on your page now.
